### PR TITLE
Feature #2395 wdir

### LINKED ISSUE #2395: Add new wind direction verification statistics for RMSE, Bias, and MAE
## Describe the New Feature ##
Wind direction verification is different from vector wind and wind speed verification. The specificity of wind direction verification is that wind direction and wind direction difference between forecast and observation need to be adjusted to (-180, +180] degree. The current VCNT and VL1L2 outputs don't support wind direction verifications of RMSE, Bias and MAE.

So here is the request list of new outputs:
1. N: the number of grid points

2. MSE (mean squared value of wind direction difference between the forecast and observed wind, this difference needs to be adjusted to (-180, +180] degree) 
DIR2_BAR=sum [(DIRf-DIRo)**2]/N, where DIRf-DIRo needs to be recalculated to  (-180, +180], DIRf is forecast wind direction and DIRo is observation wind direction.

3. Mean Error (mean value of wind direction difference between the forecast and observed wind, this difference needs to be adjusted to (-180, +180] degree) 
DIR_BAR=sum(DIRf-DIRo)/N, where DIRf-DIRo needs to be recalculated to  (-180, +180]

4. MAE (mean absolute value of wind direction difference between the forecast and observed wind, this difference needs to be adjusted to (-180, +180] degree) 
DIR_ABSBAR=sum[abs(DIRf-DIRo)]/N, where DIRf-DIRo needs to be recalculated to  (-180, +180]

Meanwhile, the outputs will allow wind speed threshold settings.

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
2 days.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [x] dtcenter/MET#2395
- [x] dtcenter/METdataio#282
- [x] dtcenter/METcalcpy#361
- [ ] dtcenter/METviewer#517

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>
- Adds 3 new [VL1L2 columns](https://met--2820.org.readthedocs.build/en/2820/Users_Guide/point-stat.html#id21):
   - DIR_ME, DIR_MAE, and DIR_MSE
- Adds 3 new [VAL1L2 columns](https://met--2820.org.readthedocs.build/en/2820/Users_Guide/point-stat.html#id22):
   - DIRA_ME, DIRA_MAE, DIRA_MSE
- Adds 12 new [VCNT columns](https://met--2820.org.readthedocs.build/en/2820/Users_Guide/point-stat.html#id23): 
   - DIR_ME, DIR_MAE, DIR_MSE, and DIR_RMSE plus _BCL and _BCU columns to store the bootstrap CI's.

Note however that the VCNT line type includes columns for bootstrap CI's, but MET *does not* actually compute them yet. Enhancing it to do so is requested in MET #963.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Manually ran `make test` and inspected the content of the new columns.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
- Meet to review the following details:
   - Review the code changes, paying particular attention to...
      - the number and names of the new output columns.
      - the computation of anomalous wind direction diffs
      - the handling of "missing" data when the direction is undefined for zero vectors. When zero vectors are encountered, the following type of warning message is written:
```
WARNING: VL1L2Info::compute_stats() -> Skipping 249 of 934 vector pairs for which the direction difference is undefined.
WARNING: Set the "wind_thresh" and "wind_logic" configuration options to exclude zero vectors.
```
   - Observe this [PR testing workflow](https://github.com/dtcenter/MET/actions/runs/7893184770) that ran without error but produced many, many diffs. Download/inspect the diffs to see that they're confined to VL1L2/VAL1L2/VCNT lines.
   - Review the description and equations added to the end of [Appendix G](https://met--2820.org.readthedocs.build/en/2820/Users_Guide/appendixG.html) for accuracy and clarity.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**
Yes, this includes updates to the Point-Stat chapter and Appendix G.

- [x] Do these changes include sufficient testing updates? **[Yes]**
New output is produced for many existing unit tests.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
A total of 84 ascii output files are modified:
   - 48 files with a `.stat` suffix.
   - 13 files with a `_vl1l2.txt` suffix.
   - 9 files with a `_val1l2.txt` suffix.
   - 14 files with a `_vcnt.txt` suffix.

- [x] Please complete this pull request review by **[Fri 2/16/24]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: src/tools/core/stat_analysis/aggr_stat_line.cc [hunk 1]
```diff
@@ -37,6 +37,8 @@
 //                    -out_thresh and -out_line_type options.
 //   017    11/10/22  Halley Gotway   MET #2339 Add SEEPS and SEEPS_MPR
 //                                      line types.
+//   018    02/13/24  Halley Gowtay   MET #2395 Add wind direction stats
```
Comment (JohnHalleyGotway) [inline src/tools/core/stat_analysis/aggr_stat_line.cc]:
```suggestion
//   018    02/13/24  Halley Gotway   MET #2395 Add wind direction stats
```

File: src/tools/core/stat_analysis/parse_stat_line.cc [hunk 2]
```diff
@@ -31,6 +31,8 @@
 //   011    01/24/20  Halley Gotway   Add RPS line type.
 //   012    11/10/22  Halley Gotway   MET #2339 Add SEEPS and SEEPS_MPR
 //                                      line types.
+//   013    02/13/24  Halley Gowtay   MET #2395 Add wind direction stats
```
Comment (JohnHalleyGotway) [inline src/tools/core/stat_analysis/parse_stat_line.cc]:
```suggestion
//   013    02/13/24  Halley Gotway   MET #2395 Add wind direction stats
```

File: src/tools/core/stat_analysis/aggr_stat_line.cc [hunk 3]
```diff
@@ -2139,6 +2141,21 @@ void aggr_mpr_wind_lines(LineDataFile &f, STATAnalysisJob &job,
...
                                    it->second.pd_v.o_na[i]*it->second.pd_v.o_na[i]);
+
+         double d_diff = angle_difference(
+                            it->second.pd_u.f_na[i], it->second.pd_v.f_na[i],
+                            it->second.pd_u.o_na[i], it->second.pd_v.o_na[i]);
+
+         if(is_bad_data(d_diff)) {
+            v_info.n_dir_undef = 1;
+         }
+	 else {
```
Comment (JohnHalleyGotway) [inline src/tools/core/stat_analysis/aggr_stat_line.cc]:
```suggestion
         else {
```

**Comment by JohnHalleyGotway** [comment]
@ericgilleland and @JohnHalleyGotway met on Feb 14, 2024 to discuss this PR. Some questions:
1. Does it make sense to add output to the VAL1L2 line type? MET #2395 requests additions to the VL1L2 and VCNT line types. I added VAL1L2 for consistency, but the question is whether or not wind direction anomaly statistics are useful to compute scientifically.
2. Per @ericgilleland, it may be possible to define parametric confidence intervals, using the von Mises distribution, for these wind direction statistics. That would require the addition of _NCL and _NCU columns in the VCNT line type. But that remains an open question. Recommend delaying for now but consider for the future.

**Review by ericgilleland — APPROVED** [review]
I approve

**Pull request merged**